### DISCUSSION TITLE
Code Generation for CSLA 5/6

### DISCUSSION THREAD

**Discussion by mar72vin** [MarimerLLC/csla#2936] [Questions]
I have been using CSLA 4.6 for quite a while and think it's finally time to start thinking about upgrading to CSLA 5/6. 

Does anyone know if there is another tool that can replace CSLAGenFork? I have been using this for years to generate business objects and stored procs based on SQL Database tables. (CSLA 4.6). 

It appears that CSLAGenFork hasn't been updated for some time. Also found this post answered by Rocky. https://github.com/MarimerLLC/csla/discussions/2050

Most of the suggested links for generators are links to codeplex which does not exist anymore or are for older versions of CSLA.

**Comment by TheCakeMonster** [comment]
Funnily enough, I do.

I've used the free version of CodeSmith for many, many years, but it runs on .NET Framework and the templates I was using were very complex to maintain, so I started to wonder what life after CodeSmith might be. In combination with work I wanted to do on streamlining the design process for agile software teams, I created [DesiGen](https://desigen-docs.dotnotstandard.com/docs) - a design and code generation tool.

This isn't just a code generator, it's a whole lot more.

The templates for generating an application using CSLA 5.5.3 are already in the box. The release of CSLA 6 yesterday has caught me on the hop as I haven't done the templates for that yet, but I'll start on them today, with a view to releasing a new version of the application next week.

You can see me demo DesiGen in a couple of videos as part of my series on agile software design on [YouTube](https://www.youtube.com/playlist?list=PLjvHiORS85l6zeDj-0WhBCCuTXYy6WKb1) - the long version with the full explanation is in videos 3 and 4, with the TL;DR demo of what is possible with custom templates in video 5.

**Reply by Chicagoan2016** [reply]
@TheCakeMonster , Could you please post a short tutorial on how to implement the business objects in Project Tracker ( in DesiGen)?
ProjectEdit.cs class should suffice.
Thanks

**Reply by TheCakeMonster** [reply]
@Chicagoan2016 OK, I'll see what I can do. I started a new job this week, but I might get a few hours on Friday.

Presumably you'd prefer it with CSLA 6, rather than with CSLA 5?

**Reply by Chicagoan2016** [reply]
Whichever version of CSLA is convenient and less time consuming for you.
Would love to see a business class (ProjectEdit) along with data Access code generated.
Thank you @TheCakeMonster .

Edit : If possible, could you please code for ADO.NET calling stored proc? nothing fancy, just a simple stored proc and ado.NET  code, we are old school and still believe in Stored procedures and 'bare-bones' ADO.NET 😀

**Reply by TheCakeMonster** [reply]
@Chicagoan2016 It would very much be my pleasure - it makes me happy to hear that stored procedures are still your preferred data access method. I too believe this to remain the most appropriate data access technology for most scenarios! As it happens, that's what my demos always include, so that makes it easier for me too!

**Reply by TheCakeMonster** [reply]
@Chicagoan2016 your tutorial awaits. ProjectTracker using DesiGen - and using CSLA 6. A full house!

Maybe order two coffees; it's quite a long one, because explaining things takes longer. Sorry.

¯\\_(ツ)_/¯

Stop, start and rewind as needs be; hopefully you can see what I'm doing from the screen capture, in combination with the audio. I have tried to explain everything - but it isn't easy to explain something to someone when you can't see the whites of their eyes, as you have no idea if they look confused or not!

https://youtu.be/VRTzHOw7uxc

**Reply by TheCakeMonster** [reply]
As well as this tutorial, I've done an installation tutorial and a couple of separate videos explaining several important security principles that we encounter along the way. I did the right thing when it comes to database security - but sometimes the _right_ thing and the _straightforward_ thing aren't the same. Oops.

https://youtu.be/jKnp7vOCEbw

https://youtu.be/-neeEBd5E2E

https://youtu.be/jQ8lNfZtiCg

**Reply by Chicagoan2016** [reply]
@TheCakeMonster , Thank you Sir, will be watching them soon! appreciate the effort.

**Comment by rockfordlhotka** [comment]
@royog approached me a while back wondering about CSLAGenFork too, suggesting that interested folks should maybe update the software. I know he looked at it and it is apparently written using now-old ASP technology, so his thought was that it probably required a full rewrite to bring it up to date.

**Comment by mtavares628** [comment]
I also had a CSLA 4 project that used CSLAGenFork and I needed to update CSLA for a rewrite of my project.  Because I had so many BOs already invested in the project, I decided to create my own [fork](https://github.com/mtavares628/CslaGenFork) of CSLAGenFork and take a stab at upgrading the templates to CSLA 5.  Mind you, I only modified the templates that I had uses for (BusinessBase, BusinessListBase, ReadOnlyBase, ReadOnlyListBase) and I only modified the C# versions.  My DAL code is also still using ADO.NET, not EF.  But, the DataPortal methods are using attributes, the DAL is being injected, and the criteria objects support multiple primitives at this point.

I'll be trying to modify the templates again for CSLA 6 very soon.  You are more than welcome to see if it meets the needs for your project, but I don't intend to fully revamp the application, I'm just modifying things to make it work for my project at this point.

**Reply by rockfordlhotka** [reply]
That's pretty nice @mtavares628 - would you be interested in merging your changes into https://github.com/marimerllc/cslagenfork ?

**Reply by TheCakeMonster** [reply]
If I didn't know better I'd think you didn't like my new and shiny generator!

😉

**Reply by mtavares628** [reply]
@rockfordlhotka sure I can submit a PR, as long as you are comfortable with it not being a finished product for all templates.  I pretty much copied the Csla 4 template folders, and then adjusted them for the BOs that I needed, so there are a number of files in that CSLA5 templates folder that are still CSLA 4.  If that's okay with you, then I will create the PR.

**Reply by rockfordlhotka** [reply]
> If I didn't know better I'd think you didn't like my new and shiny generator!
> 
> 😉

Not at all! 😊

Folks with an investment in existing templates may prefer to stick with the tools they already use though.

**Reply by rockfordlhotka** [reply]
> @rockfordlhotka sure I can submit a PR, as long as you are comfortable with it not being a finished product for all templates.  I pretty much copied the Csla 4 template folders, and then adjusted them for the BOs that I needed, so there are a number of files in that CSLA5 templates folder that are still CSLA 4.  If that's okay with you, then I will create the PR.

That's totally fine with me.

**Reply by mtavares628** [reply]
@rockfordlhotka I did submit that [PR](https://github.com/MarimerLLC/CslaGenFork/pull/1). Not sure if you saw it or not.

**Reply by rockfordlhotka** [reply]
Thank you, I did not. I've turned on all notifications for that repo, so I will in the future.

I accepted the PR - that's awesome!

I also did the rename of master to main, as I think that's important.

**Comment by Chicagoan2016** [comment]
I would love to see some kind of short tutorial for Cslagenfork!

**Reply by mtavares628** [reply]
Unfortunately, the closest thing you are going to find is this [CslaGenWiki](http://cslagen.wiki-site.com/index.php/Main_Page)

And here's info from the original CslaGenFork site on where to get CodeSmith and set it up: [CodeSmith Setup Guide](https://github.com/CslaGenFork/CslaGenFork/wiki/CodeSmith-Setup-Guide)
